lms/djangoapps/lti_provider/signature_validator.py

```diff
@@ -0,0 +1,182 @@
+        """
+        Verify that the nonce value that accompanies the OAuth signature is
+        valid. This method is concerned only with the structure of the nonce;
+        the validate_timestamp_and_nonce method will check that the nonce has
+        not been used within the specified time frame. This method signature is
+        required by the oauthlib library.
+
+        :return: True if the OAuth nonce is valid, or False if it is not.
+        """
+        return nonce is not None and 0 < len(nonce) <= 64
+
+    def validate_timestamp_and_nonce(self, client_key, timestamp, nonce,
+                                     request, request_token=None,
+                                     access_token=None):
+        """
+        Verify that the request is not too old (according to the timestamp), and
+        that the nonce value has not been used already within the period of time
+        in which the timestamp marks a request as valid. This method signature
+        is required by the oauthlib library.
+
+        :return: True if the OAuth nonce and timestamp are valid, False if they
+        are not.
+        """
+        return True
+
+    def validate_client_key(self, client_key, request):
+        """
+        Ensure that the client key supplied with the LTI launch is on that has
+        been generated by our platform, and that it has an associated client
+        secret.
+
+        :return: True if the key is valid, False if it is not.
+        """
+        return LtiConsumer.objects.filter(key=client_key).count() == 1
+
+    def get_client_secret(self, client_key, request):
+        """
+        Fetch the client secret from the database. This method signature is
+        required by the oauthlib library.
+
+        :return: the client secret that corresponds to the supplied key if
+        present, or None if the key does not exist in the database.
+        """
+        try:
+            return LtiConsumer.objects.get(key=client_key).secret
+        except ObjectDoesNotExist:
+            return None
+
+    def verify(self, request):
+        """
+        Check the OAuth signature on a request. This method uses the
+        SignatureEndpoint class in the oauthlib library that in turn calls back
+        to the other methods in this class.
+
+        :param request: the HttpRequest object to be verified
+        :return: True if the signature matches, False if it does not.
+        """
+
+        method = unicode(request.method)
+        url = request.build_absolute_uri()
+        body = request.body
+
+        # The oauthlib library assumes that headers are passed directly from the
+        # request, but Django mangles them into its own format. The only header
+        # that the library requires (for now) is 'Content-Type', so we
+        # reconstruct just that one.
+        headers = {"Content-Type": request.META['CONTENT_TYPE']}
+
+        result, _ = self.endpoint.validate_request(url, method, body, headers)
```

Comment:
`_` is commonly aliased to `ugettext` in our python files, so I request that people use `__` or `_dummy` as placeholder variables, as to not inadvertantly overload the meaning of `_` as gettext.